### ISSUE TITLE
[red-knot] `T & ~X` should be assignable to `T`

### ISSUE THREAD

**Issue by carljm** [astral-sh/ruff#17364]
[This](https://types.ruff.rs/327f3b66-7d85-452b-9af7-66c1ede9913f) does not type check, but it should:

```py
def union_param[T](x: T | None) -> T:
    if x is None:
        raise ValueError("don't give me that none")
    return x  # bogus error: expected T, found T & ~None
```

[The equivalent](https://types.ruff.rs/9e7dc33b-168b-44c3-b06d-e5ce004523c2) where we replace `T` with a concrete type, does type check:

```py
def union_param(x: int | None) -> int:
    if x is None:
        raise ValueError("don't give me that none")
    return x
```

Placing this in alpha milestone because I've run into it already several times, I think it will be a common source of false positives in generic functions.
